src/Backend/Haskell.idr

```diff
@@ -346,48 +400,68 @@ freshEnvWithTerms : (prefix : String) -> Vect n (HsType, HsTerm)
 freshEnvWithTerms prefix = map (\ x => (x, HsTermVar (prefix ++ uppercase (hsTypeName x)))) freshEnv
 
 ||| Generate a Haskell type from a `TDef`.
-makeType : Vect n HsType -> TDef n -> HsType
-makeType _    T0          = HsVoid
-makeType _    T1          = HsUnit
-makeType e    (TSum xs)   = foldr1' (\t1,t2 => HsSum t1 t2) (map (assert_total $ makeType e) xs)
-makeType e    (TProd xs)  = HsTuple $ map (assert_total $ makeType e) xs
-makeType e    (TVar v)    = Vect.index v e
-makeType e td@(TMu cases) = HsParam (nameMu cases) $ getUsedVars e td
-makeType e    (TApp f xs) = HsParam (name f) (map (assert_total $ makeType e) xs)
+makeType : Vect n HsType -> TDef n -> HaskellLookupM HsType
+makeType _    T0          = pure HsVoid
+makeType _    T1          = pure HsUnit
+makeType e    (TSum xs)   = do ys <- traverseEffect (assert_total $ makeType e) xs
+                               pure $ foldr1' HsSum ys
+makeType e    (TProd xs)  = do ys <- traverseEffect (assert_total $ makeType e) xs
+                               pure $ HsTuple ys
+makeType e    (TVar v)    = pure $ Vect.index v e
+makeType e td@(TMu cases) = pure $ HsParam (nameMu cases) $ getUsedVars e td
+makeType e    (TApp f xs) = do ys <- (traverseEffect (assert_total (makeType e)) xs)
+                               pure $ HsParam (name f) ys
+makeType e    (TRef n)    = case lookup n !('Spec :- get) of
+                              Just t => pure t
+                              Nothing => raise $ RefNotFound n
 
 ||| Generate a Haskell type from a `TNamed`.
-makeType' : Vect n HsType -> TNamed n -> HsType
+makeType' : Vect n HsType -> TNamed n -> HaskellLookupM HsType
 makeType' e (TName ""   body) = makeType e body --escape hatch; used e.g. for all non-TApp dependencies of a TApp
-makeType' e (TName name body) = HsParam name $ getUsedVars e body
+makeType' e (TName name body) = pure $ HsParam name $ getUsedVars e body
+
+ifNotPresent' : Name -> HaskellDefM (List Haskell) -> HaskellDefM (List Haskell)
```

Comment:
Is this a specialized version of `Backend.Utils.ifNotPresent` to make life easier for the typechecker, or is there something deeper going on? Either way might be useful to add a comment on why it's here.

Reply:
It's an artefact of previous attempts, I removed it and replaced it with `Backend.Utils.ifNotPresent`